includes/cli/class-amp-cli-validation-command.php

```diff
@@ -314,6 +314,69 @@ public function reset( $args, $assoc_args ) {
 		WP_CLI::success( 'All AMP validation data has been removed.' );
 	}
 
+	/**
+	 * Generate the authorization nonce needed for a validate request.
+	 *
+	 * @subcommand generate-nonce
+	 * @alias nonce
+	 */
+	public function generate_nonce() {
+		WP_CLI::line( AMP_Validation_Manager::get_amp_validate_nonce() );
+	}
+
+	/**
+	 * Get the validation results for a given URL.
+	 *
+	 * The results are returned in JSON format.
+	 *
+	 * ## OPTIONS
+	 *
+	 * <url>
+	 * : The URL to check. The host name need not be included. The URL must be local to this WordPress install.
+	 *
+	 * ## EXAMPLES
+	 *
+	 *     wp amp validation check-url /about/
+	 *     wp amp validation check-url $( wp option get home )/?p=1
+	 *
+	 * @subcommand check-url
+	 * @alias check
+	 *
+	 * @param array $args Args.
+	 */
+	public function check_url( $args ) {
+		list( $url ) = $args;
+
+		$host            = wp_parse_url( $url, PHP_URL_HOST );
+		$parsed_home_url = wp_parse_url( home_url( '/' ) );
+
+		if ( $host && $host !== $parsed_home_url['host'] ) {
+			WP_CLI::error(
+				sprintf(
+					/* translators: %1$s is the expected host, %2$s is the actual host */
+					__( 'Supplied URL must be for this WordPress install. Expected host "%1$s" but provided is "%2$s".', 'amp' ),
+					$parsed_home_url['host'],
+					$host
+				)
+			);
+		}
+
+		if ( ! $host ) {
+			$origin = $parsed_home_url['scheme'] . '://' . $parsed_home_url['host'];
+			if ( ! empty( $parsed_home_url['port'] ) ) {
+				$origin .= ':' . $parsed_home_url['port'];
+			}
+			$url = $origin . '/' . ltrim( $url, '/' );
+		}
+
+		$result = AMP_Validation_Manager::validate_url( $url );
+		if ( $result instanceof WP_Error ) {
+			WP_CLI::error( $result );
+		}
+
+		print wp_json_encode( AMP_Validation_Manager::validate_url( $url ), JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES );
```

Comment:
Would it be better to use `WP_CLI::line()` here?

Reply:
`WP_CLI::log()` should be used here. `WP_CLI::line()` is a legacy compat method.

Reply:
No, forget my rambling. You are correct, `WP_CLI::line()` should be used here, as this output should not be suppressed when using the `--quiet` flag.

Reply:
Actually, given the input, the command should support a `--format` flag and the output should be processed via `WP_CLI\Utils\format_items()` instead.

This way it would default to the normal WP-CLI tabular output, but you could request JSON if you wanted to.

Reply:
I don't think the data would work well in a tabular format. Would it? It looks like this:

```json
{
    "results": [
        {
            "error": {
                "node_name": "script",
                "parent_name": "div",
                "code": "DISALLOWED_TAG",
                "type": "js_error",
                "node_attributes": [],
                "text": "document.write(\"BAD!\");",
                "sources": [
                    {
                        "hook": "the_content",
                        "filter": true,
                        "post_id": 2531,
                        "post_type": "post",
                        "sources": [
                            {
                                "type": "core",
                                "name": "wp-includes",
                                "file": "class-wp-embed.php",
                                "line": 59,
                                "function": "WP_Embed::run_shortcode"
                            },
                            {
                                "type": "core",
                                "name": "wp-includes",
                                "file": "class-wp-embed.php",
                                "line": 404,
                                "function": "WP_Embed::autoembed"
                            },
                            {
                                "type": "core",
                                "name": "wp-includes",
                                "file": "blocks.php",
                                "line": 532,
                                "function": "do_blocks"
                            },
                            {
                                "type": "core",
                                "name": "wp-includes",
                                "file": "formatting.php",
                                "line": 51,
                                "function": "wptexturize"
                            },
                            {
                                "type": "core",
                                "name": "wp-includes",
                                "file": "formatting.php",
                                "line": 455,
                                "function": "wpautop"
                            },
                            {
                                "type": "core",
                                "name": "wp-includes",
                                "file": "formatting.php",
                                "line": 838,
                                "function": "shortcode_unautop"
                            },
                            {
                                "type": "core",
                                "name": "wp-includes",
                                "file": "post-template.php",
                                "line": 1649,
                                "function": "prepend_attachment"
                            },
                            {
                                "type": "core",
                                "name": "wp-includes",
                                "file": "media.php",
                                "line": 1432,
                                "function": "wp_make_content_images_responsive"
                            },
                            {
                                "type": "core",
                                "name": "wp-includes",
                                "file": "formatting.php",
                                "line": 5726,
                                "function": "wp_staticize_emoji"
                            },
                            {
                                "type": "core",
                                "name": "wp-includes",
                                "file": "formatting.php",
                                "line": 5359,
                                "function": "capital_P_dangit"
                            },
                            {
                                "type": "core",
                                "name": "wp-includes",
                                "file": "shortcodes.php",
                                "line": 177,
                                "function": "do_shortcode"
                            },
                            {
                                "type": "core",
                                "name": "wp-includes",
                                "file": "formatting.php",
                                "line": 3335,
                                "function": "convert_smilies"
                            }
                        ]
                    },
                    {
                        "block_name": "core/shortcode",
                        "post_id": 2531,
                        "block_content_index": 0,
                        "type": "plugin",
                        "name": "gutenberg",
                        "file": "build/block-library/blocks/shortcode.php",
                        "line": 16,
                        "function": "gutenberg_render_block_core_shortcode"
                    },
                    {
                        "type": "plugin",
                        "name": "bad-shortcode.php",
                        "file": "bad-shortcode.php",
                        "line": 7,
                        "function": "Bad\\shortcode",
                        "shortcode": "bad"
                    }
                ]
            },
            "sanitized": true
        },
        {
            "error": {
                "node_name": "script",
                "parent_name": "div",
                "code": "DISALLOWED_TAG",
                "type": "js_error",
                "node_attributes": [],
                "text": "badness",
                "sources": [
                    {
                        "hook": "the_content",
                        "filter": true,
                        "post_id": 2531,
                        "post_type": "post",
                        "sources": [
                            {
                                "type": "core",
                                "name": "wp-includes",
                                "file": "class-wp-embed.php",
                                "line": 59,
                                "function": "WP_Embed::run_shortcode"
                            },
                            {
                                "type": "core",
                                "name": "wp-includes",
                                "file": "class-wp-embed.php",
                                "line": 404,
                                "function": "WP_Embed::autoembed"
                            },
                            {
                                "type": "core",
                                "name": "wp-includes",
                                "file": "blocks.php",
                                "line": 532,
                                "function": "do_blocks"
                            },
                            {
                                "type": "core",
                                "name": "wp-includes",
                                "file": "formatting.php",
                                "line": 51,
                                "function": "wptexturize"
                            },
                            {
                                "type": "core",
                                "name": "wp-includes",
                                "file": "formatting.php",
                                "line": 455,
                                "function": "wpautop"
                            },
                            {
                                "type": "core",
                                "name": "wp-includes",
                                "file": "formatting.php",
                                "line": 838,
                                "function": "shortcode_unautop"
                            },
                            {
                                "type": "core",
                                "name": "wp-includes",
                                "file": "post-template.php",
                                "line": 1649,
                                "function": "prepend_attachment"
                            },
                            {
                                "type": "core",
                                "name": "wp-includes",
                                "file": "media.php",
                                "line": 1432,
                                "function": "wp_make_content_images_responsive"
                            },
                            {
                                "type": "core",
                                "name": "wp-includes",
                                "file": "formatting.php",
                                "line": 5726,
                                "function": "wp_staticize_emoji"
                            },
                            {
                                "type": "core",
                                "name": "wp-includes",
                                "file": "formatting.php",
                                "line": 5359,
                                "function": "capital_P_dangit"
                            },
                            {
                                "type": "core",
                                "name": "wp-includes",
                                "file": "shortcodes.php",
                                "line": 177,
                                "function": "do_shortcode"
                            },
                            {
                                "type": "core",
                                "name": "wp-includes",
                                "file": "formatting.php",
                                "line": 3335,
                                "function": "convert_smilies"
                            }
                        ]
                    },
                    {
                        "block_name": "core/html",
                        "post_id": 2531,
                        "block_content_index": 1
                    }
                ]
            },
            "sanitized": true
        },
        {
            "error": {
                "node_name": "script",
                "parent_name": "body",
                "code": "DISALLOWED_TAG",
                "type": "js_error",
                "node_attributes": [],
                "text": "document.write( 'I am invalid!!' );",
                "sources": [
                    {
                        "type": "plugin",
                        "name": "print-invaliud-footer.php",
                        "file": "print-invaliud-footer.php",
                        "line": 8,
                        "function": "{closure}",
                        "hook": "wp_footer",
                        "priority": 10
                    }
                ]
            },
            "sanitized": true
        }
    ],
    "queried_object": {
        "id": 2531,
        "type": "post"
    },
    "url": "https://wordpressdev.lndo.site/2019/12/05/bad-shortcode/?amp"
}
```

Reply:
I've improved the command to make use of the standard WP-CLI data offerings. This now includes JSON, YAML and table formatting, counting, exit codes. But also filtering on a given field and selecting which fields. For table output, I improved the formatting so it is useful as the human-readable format.

Here's an example of table output:

<img width="1675" alt="Image 2019-12-14 at 2 25 21 AM" src="https://user-images.githubusercontent.com/83631/70841252-4b9fc900-1e19-11ea-9b30-1b9a0ce90774.png">

To filter, you can do something like this:
```
wp amp validation check-url example.org --type=css_error
```

What do you think about that format, it seems more useful for human consumption than JSON, doesn't it?

Should I commit this into this current PR, or create a separate one?

Reply:
A problem with that is the fields vary by error type. Maybe you accounted for this. Doing this in a separate PR makes sense to me.

The table does look nice. But at the same time, the admin screen has this information already presented in an HTML table. I think that perhaps this command should in addition to obtaining the validation results, also store them and then provide the URL to the validated URL screen via a message to `STDERR`.